### ISSUE TITLE
How to start a "play online" server

### ISSUE THREAD

**Issue by joelpurra** [farin/JCloisterZone#87]
I haven't seen any instructions on how to start a server for the "play online" feature. Are there any?

**Comment by farin** [comment]
There is no start. Play server is play.jcloisterzone.com and it's running.

**Comment by joelpurra** [comment]
@farin: will it be open sourced?

**Comment by farin** [comment]
@joelpurra It's not on list now

**Comment by joelpurra** [comment]
@farin: it would be lovely to have it as open source! JCZ is already licensed under the [GNU Affero General Public License, Version 3 (AGPL-3.0)](https://www.gnu.org/licenses/agpl-3.0.html), and that is perfect for a server component for an open source client.

**Comment by farin** [comment]
I want to keep control of server for now. Because makes me not much sense shatter such feature. Centralized service can provide more features like players rankings etc. But I can change mind in future (jcz itself was closet at early begining too)

I woudl like to add simple standalone server support to jcz itself. So you will be able to run it without ui on your own server if you want private play, bypass nat or whatever you want. This is easy and should appear in 3.1 too.

**Comment by joelpurra** [comment]
@farin: as the distributor of the pre-built packages containing the default domain settings for the play online server, you are already in control. Releasing the server source is a matter of openness - and that applies to game statistics, player ranking and other communication related code as well. Plus the AGPL license prevents forking.

Seeing that people work to improve the JCZ client together, it doesn't make much sense to keep the server private.

**Comment by farin** [comment]
@joelpurra Maybe you're right. I will reconsider when server will be stabilized and cleaned.

**Comment by joelpurra** [comment]
@farin: looking forward to it =)

**Comment by joelpurra** [comment]
Oh, and the reason I was asking was because when we tried a four-player "play online" game (instead of "connect", which works well), the three "clients" all got connection timeouts during after a minute or two in the lobby. We didn't try to start the game after that.

**Comment by farin** [comment]
@joelpurra could you described how exactly looks connection timeout? you were disconnected from lobby? It's tru that current impl is not precisely concurrent (only one websocket worker based on greenthreads), but four client's shouldn't be problem. I personally tried 5 client game and it looked all fine.

**Comment by joelpurra** [comment]
@farin: I hosted a new game in "play online". Three additional players connect right away, and enter the lobby. After a while two of them asked, over third-party voice chat, what was going on with the game. Turns out they had been disconnected. The fourth player was still in the room, as seen by both host and client. Within a few seconds the fourth player was also disconnected, as confirmed over voice chat.

No debug information was collected. We had already played multiple games before, using the normal "connect." All players were located in Sweden. The host and one of the the players are behind the same NAT, but the player was disconnected nonetheless.

It could be the distance to Sweden affecting message delays? If the client expects a reply from the host's game server, response times are probably double that of a normal connection between the client and `play.jcloisterzone.com`, as messages pass the intermediate server. This is on top of `play.jcloisterzone.com` presumably not being located in Sweden, which means ping times can be much higher than intra-country games.
